xtp/src/libxtp/pmlocalization.cc

```diff
@@ -0,0 +1,249 @@
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * Reference- A fast intrinsic localization procedure applicable for ab initio
+ * and semiempirical linear combination of atomic orbital wave functions J.
+ * Chem. Phys. 90, 4916 (1989); https://doi.org/10.1063/1.456588 János Pipek and
+ * Paul G. Mezey
+ */
+
+#include "votca/xtp/pmlocalization.h"
+#include "votca/xtp/aomatrix.h"
+#include <limits>
+
+namespace votca {
+namespace xtp {
+void PMLocalization::computePML(Orbitals &orbitals) {
+  occupied_orbitals = orbitals.MOs().eigenvectors().leftCols(
+      orbitals.getNumberOfAlphaElectrons());
+  aobasis = orbitals.getDftBasis();
+  AOOverlap overlap;
+  overlap.Fill(aobasis);
+  overlap_ = overlap.Matrix();
+
+  XTP_LOG(Log::error, log_) << std::flush;
+  XTP_LOG(Log::error, log_)
+      << TimeStamp() << " Starting localization of orbitals" << std::flush;
+
+  // determine initial cost_function
+  initial_cost_function();
+
+  Index iteration = 1;
+  Index maxrow, maxcol;
+  double max_cost = PM_cost_function_.maxCoeff(&maxrow, &maxcol);
+
+  bool not_converged = true;
+  while (not_converged && iteration < nrOfIterations_) {
+
+    XTP_LOG(Log::info, log_) << "Iteration: " << iteration << std::flush;
+
+    XTP_LOG(Log::info, log_)
+        << "Orbitals to be changed: " << maxrow << " " << maxcol << std::flush;
+
+    Eigen::MatrixX2d max_orbs(occupied_orbitals.rows(), 2);
+    max_orbs << occupied_orbitals.col(maxrow), occupied_orbitals.col(maxcol);
+    Eigen::MatrixX2d rotated_orbs = rotateorbitals(max_orbs, maxrow, maxcol);
+    occupied_orbitals.col(maxrow) = rotated_orbs.col(0);
+    occupied_orbitals.col(maxcol) = rotated_orbs.col(1);
+
+    update_cost_function(maxrow, maxcol);
+
+    // check for convergence
+    max_cost = PM_cost_function_.maxCoeff(&maxrow, &maxcol);
+
+    if (max_cost < convergence_limit_) not_converged = false;
+
+    XTP_LOG(Log::info, log_)
+        << "maximum of penalty function: " << max_cost << std::flush;
+
+    iteration++;
+  }
+  XTP_LOG(Log::error, log_) << TimeStamp() << " Orbitals localized after "
+                            << iteration + 1 << " iterations" << std::flush;
+  orbitals.setPMLocalizedOrbital(occupied_orbitals);
+}
+
+// Function to rotate the 2 maximum orbitals (s and t)
+Eigen::MatrixX2d PMLocalization::rotateorbitals(const Eigen::MatrixX2d &maxorbs,
+                                                const Index s, const Index t) {
+  const double gamma =
+      0.25 * asin(B(s, t) / sqrt((A(s, t) * A(s, t)) + (B(s, t) * B(s, t))));
```

Comment:
add std:: in front of the functions, otherwise you will get into weird behaviour